File: Data Structures & Algorithms/partition-to-k-equal-sum-subsets/submission-6.cpp
#include <bits/stdc++.h>
class Solution {
public:
vector<int> nums;
int t=0,k;
    bool canPartitionKSubsets(vector<int>& nums, int k) {
      this->nums=nums;
      this->k=k;
      vector<int> sums(k,0);
      sort(nums.begin(),nums.end(),greater<int>());

    for(int i: nums)
    t+=i;
    if(t%k!=0)
    return false;
    t=t/k;
    return backTrack(sums,0);
    }
    //j is the index to iterate through nums
    bool backTrack(vector<int>& sums,int j){
        if(j==nums.size()) return true;
       for(int i=0;i<k;i++){
        int x=nums[j];
        if(sums[i] + x>t) continue;
        sums[i]+=x;
        if(backTrack(sums,j+1)) return true;
        //backTrack(sums,j+1) is false when nums[j] doesnt go into any oh=f teh buckets
        sums[i]-=x;
        if(sums[i]==0) break;
       }
        return false;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-6.java
class Solution {
    public int search(int[] nums, int target) {
        int l=0,r=nums.length-1,m;
        if(nums[l]==target)
        return l;
        if(nums[r]==target)
        return r;


        while(l<=r){
            m=(l+r)/2;

            if(nums[m]==target)
            return m;

            if(nums[l]<=nums[m]){
             if( target < nums[m] && target >= nums[l])
            r=m-1;
            else
            l=m+1;
            }
            else{
            if(nums[r]>=nums[m] && target>nums[m] && target<=nums[r])
            l=m+1;
            else
            r=m-1;
            }
        }

        return -1;
    }
}
